### ISSUE TITLE
Create Docker container for easier distribution

### ISSUE THREAD

**Issue by thomashopf** [debbiemarkslab/EVcouplings#74]
* including all software dependencies, as far as possible
* think about how to best handle databases
* pay attention to licenses, some of the used tools are under GPL

**Comment by sacdallago** [comment]
You would usually not dockerize databases.

What you can do is create a separate container for the database and then create a docker-compose script that pulls both (or more) containers at once. I've done it multiple times, latest in this project: https://github.com/efktr/docker-compose

I will take care of this.

What is the problem with licenses? If the Readme of this repo acknowledges the tools (and complies with their licenses), then you should be good to go :)

**Comment by b-schubert** [comment]
> What you can do is create a separate container for the database and then create a docker-compose script that pulls both (or more) containers at once. I've done it multiple times, latest in this project: https://github.com/efktr/docker-compose

Well isn't this a dockerization of a database then ;-)

> What is the problem with licenses? 

As Thomas indicated, some of the dependencies are released under GPL, which means the docker container must also be released under a GPL license. 

Not a big issue as it would not influence the licensing of the API package.

**Comment by thomashopf** [comment]
Yep the first thing would be to check what needs to be done re the tools with special licenses, in particular CNS. psipred needs checking too.

**Comment by sacdallago** [comment]
@b-schubert 
> Well isn't this a dockerization of a database then ;-)

Not entirely :D the data is still stored separately from the container. Just the database container is created, and on first run pulls the data. This means slow first boot, but small containers and sort of maintaining the whole paradigm

@thomashopf I see. This is a separate discussion. I'll open an issue for it

**Comment by nselem** [comment]
I am creating a [docker container](https://github.com/nselem/evcouplings)   
It works now trough align, compare, couplings, and mutate. I'm still having troubles with fold because of cns installation.
   
For now I solved the cns license problem with an ADD, everybody must have its own copy of cns_solve:   
`ADD cns_solve_1.21_all-mp.tar.gz /opt`  
And first time databases command must be called `evcouplings_dbupdate`  

Please could you help me with cns ? I got this error:
evcouplings.utils.system.ExternalToolError: Call to external tool failed and did not return: /opt/cns_solve_1.21/bin

**Comment by nselem** [comment]
Hello @sacdallago 

Thank you for your answer, well if you help me I can develop these ideas, because the installation is really complex by now. Sounds great to me to split the job three containers. In my case the database command were run only once inside my big docker container, so I didn't actually had python or EVcouplings installed outside the container, but anyway I think you are right and it is better to have another container. 

Back to my issue, I have decompressed cns in the directory /opt, the content is the following:     
`ls /opt/cns_solve_1.21/ `   
`Changes  License  Makefile  Readme  bin  cns_solve_env	doc  helplib  inputs  instlib  libraries  modules  source  test  utils`  

I have seen that your config example has this line:   
`cns: /n/groups/marks/pipelines/evcouplings/software/cns_solve_1.21/intel-x86_64bit-linux/bin/cns`  

then I have tried   
`cns: /opt/cns_solve_1.21/bin/`   
`cns: /opt/cns_solve_1.21/source/`
`cns: /opt/cns_solve_1.21`   
  
but it does not work   and this is my bin content:  
`apply_patch  cns_header  cns_transfer  create_patch  getarch  maketar	   objects    run_inputs  syntax_convert  update_version   www_cns_remove  www_server_remove
cns_edit     cns_info	 cns_web       depend	     install  modify_path  run_diffs  run_tests   update_patch	  windows_install  www_cns_setup   www_server_setup`   

I just decompress and didnt run cns_solve_env command or the make install, because tutorials said that it is not necessary. 
This is the one tha I used:  
/home/strain/Descargas/cns_solve_1.21_all-mp.tar.gz  
but now I'm going to try this:  
/home/strain/Descargas/cns_solve_1.21_all_intel-mac-mp.tar.gz

**Comment by nselem** [comment]
ADD command from docker decompress automatically tar.gz files in the selected directory (/opt in this case)

**Comment by nselem** [comment]
@sacdallago 
Seems like I got [EVcouplings in three docker containers](https://github.com/nselem/evcouplings)  exposing the volumes. It worked once.  I will try again to cleanly follow the tutorial from the beginning and I let you know.  I just changed the cns version to cns_solve_1.21_all_intel-mac-mp.tar.gz.  
Thank you.

**Comment by thomashopf** [comment]
@nselem: Welcome and thanks a lot for contributing!

Re your CNS problems, the pipeline expects the path to the actual executable (e.g. cns_solve_1.21/intel-x86_64bit-linux/bin/cns) for the target platform. Can't remember the details but I seem to recall that was the only way to get CNS to work reliably without the configuration shell scripts.

#### Design goals:

* Ideally, for users this should be a one-stop shop installation process (fetch or build, then download/create databases)... I suppose this should be possible with docker-compose.

* I think it would be best if job output directories are outside the container (i.e. some directory is mapped as a volume *into* the container) rather than having using mapping some volume inside the container... this way round is more intuitive

* If we cannot sort out the licensing issues for the tools in the fold stage, making the folding stage an optional dependency (I like the ADD solution) could be a way out - if the tools aren't there, the fold stage gets disabled in the default config used by the container.

#### External tools and licenses (other than plmc):
* HMMER: As of the latest release this is BSD-licensed instead of GPL. Can also be installed using conda using ``` conda install -c bioconda hmmer```, so if this goes into container that has (mini)conda, that would save building from source.
* HHsuite (for hhfilter): GPL-licensed, so this drives the license of the Docker container to GPL
* CNS: most problematic one
* PSIPRED: license sounds largely unproblematic on first read but needs additional checking
* maxcluster: website currently down... don't think it had a license

So in summary the most difficult tools license-wise are all in the fold stage. So we would either need to ask the authors for permission, or we make them optional or mandatory installs using ADD.

**Comment by sacdallago** [comment]
Hi @nselem , thanks for the work and update. I'll have a look at the dockerfiles in the next couple days, most likely during Christmas. Will try to leave comments on your repo. Also: thanks for shining light on the `ADD` instruction, pretty cool stuff with the auto unzip :) As @thomashopf identified, it seems that your problem is the absolute path to the binary in the config file. But, from your `ls` of bin, I see there's no clear `cns` binary either. I did some digging. The binary that you want is the following:

```
/.../cns_solve_1.21/intel-x86_64bit-linux/source/cns_solve-0901242058.exe
```

you may be able to pass also the following:

```
/.../cns_solve_1.21/intel-x86_64bit-linux/bin/cns
```

but that is a symlink to what above. Notice that we are in the `cns_solve_1.21/intel-x86_64bit-linux` directory, **not** in the `cns_solve_1.21/bin` directory.

Also: don't worry about the extension (`.exe`), it's not a MS file.

@thomashopf all that you said is right.

The reason why I would dedicate an entire image for dependencies is because the container produced from that image could be kept "offline". The other two images can easily be published to docker hub. Following your suggestions, we might want to also deploy an image of the dependencies image excluding fold dependencies.

Unfortunately, there's no way of using relative paths (to external, local fs) directories from within the container (which is dangerous, anyway) without including that instruction _somewhere_. The `VOLUME` instruction inside the dockerfile, to me, is just a hint to the users of stuff that needs to be taken care of (e.g.: map this container volume to local fs, or nfs for job results). Even if you define a volume inside the container, on `docker run` that volume needs to be mapped to something (the "way around" that you propose is always necessary! Just that with VOLUMES inside the dockerfile you ship the image giving that location a "logical" name). Otherwise the volume will just be some space in the container that you can see with `docker info`. For the two images with databases and dependencies, the `VOLUMES` are necessary in combination with docker-compose to tell the _compute_ container where to get the dependencies from. Hope this makes sense :)

**Comment by sacdallago** [comment]
Thanks to @nselem for his implementation, I think we have a working solution for this [here](https://github.com/nselem/evcouplings)
